Solar System

MrChrisW opened the pull request:
# Solar System IA :rocket: :new_moon: 

**What does your Instant Answer do?**
Return various attributes of objects in the solar system

**What is the data source for your Instant Answer? (Provide a link if possible)**
NASA: http://solarsystem.nasa.gov/planets/index.cfm

**What are some example queries that trigger this Instant Answer?**

##### Supported Objects
*earth, jupiter, mars, mercury, neptun, saturn, uranus, venus, sun*

##### Supported Attributes
* radius, volume, mass, surface area
* size => radius
* area => surface area

##### Example Triggers
* `<attribute> of <planet>`
* `what is the <attribute> of <planet>`

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Astronomers :smile:

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**

** New screenshots see [below](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/939#issuecomment-73118150)*

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [X] Google Chrome
        [X] Firefox

    - Windows 7
        [X] Google Chrome
        [X] Firefox
        [X] IE 8

    - Windows XP
        [X] IE 8

	- Linux
        [X] Chromium
        [X] Firefox
```

------
> IA Page https://duck.co/ia/view/solar_system

MrChrisW commented on lib/DDG/Goodie/Planets.pm:
```diff
@@ -0,0 +1,109 @@
+
+  # Switch to imperial for non-metric countries
+  # https://en.wikipedia.org/wiki/Metrication
+  if ($loc->country_code =~ m/UK|US|MM|LR/i) {
```
I'm going to get in trouble for this conditional statement :laughing: The only country I'm not sure about is the UK (they still have road signs in imperial)

Ultimately it may be best to show both like **3,928 miles (6,321 kilometers)** and swap the primary number based on the location.

moollaza replied:
@MrChrisW actually we now have a Setting that users can set to signify their Units of Measure preference. We should switch it on the frontend according to that, like we do for Forecast

moollaza replied:
https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/forecast/forecast.js#L377

MrChrisW replied:
@moollaza Can I access this preference in the back-end (perl)? It's a little difficult to add front end JS to a Goodie, however it is possible.

mintsoft replied:
For the record; the UK is metric first and imperial second

MrChrisW replied:
@mintsoft Thanks for clearing that up, I found it quite confusing that the country has adopted a partial metrication policy. In Australia we used British Imperial units back in the 70s (before my time) :laughing: 

So I'd like to clarify the best way to proceed:
 1. Switching units based on users preferences (not sure this is accessible without using JS)

 2. Switching units based on location (currently in use)

 3. Not switching units, all units as SI (modern metric)

Let me know your thoughts.

mintsoft replied:
I'd say SI and sod the Americans; however I'm sure they have other opinions :)

MrChrisW replied:
@mintsoft I'm inclined to agree good sir, :tea: The fact the NASA website also lists the metric values first further justifies it. http://solarsystem.nasa.gov/planets/profile.cfm?Object=Mercury&Display=Facts

I'll await feedback :smile:

MrChrisW commented on share/goodie/planets/planets.yml:
```diff
@@ -0,0 +1,154 @@
+---
+sun:
+  radius: 695,508 km
+  radius_imperial: 432,168.6 miles
```
There's some incontinences with naming the units. Should `miles` be `mi` or should `km` be kilometers` ?

mintsoft replied:
Personally I'd keep everything SI (i.e. Metres)

MrChrisW commented on lib/DDG/Goodie/Planets.pm:
```diff
@@ -0,0 +1,109 @@
+    neptune => goodie_img_tag({filename => 'images/Neptune.svg',height => 48, width => 48,}),
+    saturn => goodie_img_tag({filename => 'images/Saturn.svg',height => 48, width => 48,}),
+    uranus => goodie_img_tag({filename => 'images/Uranus.svg',height => 48, width => 48,}),
+    venus => goodie_img_tag({filename => 'images/Venus.svg',height => 48, width => 48,})
```
These icons require [attribution](http://thenounproject.com/kevinwhitexD/) so we may want to change them for an internal design or completely remove images. I'll wait and see what happens when @abeyang gets his hands on this. :laughing:

MrChrisW commented:
IE8 can only do Base64 URI images up to 32kb in size. I'm unsure about other versions of IE? 

![selection_075](https://cloud.githubusercontent.com/assets/5282264/5818732/ce9f255c-a0f2-11e4-9316-1fb87d6e6ab4.png)

[SunInfo](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/SunInfo.pm) is also broken on IE8.

moollaza commented:
@MrChrisW this is looking pretty good so far!

Regarding SunInfo there's some discussion about how to fix it. Really the problem is the Goodie::ImageRole implementation. We have a better way of handling SVG's on the frontend via JS/CSS Classes. See https://github.com/duckduckgo/zeroclickinfo-goodies/issues/770#issuecomment-67340979

MrChrisW commented:
Thanks @moollaza, hopefully the design can be improved upon. 

I'll use the CSS hack to workaround the SVG issues in IE, provided we still include images!

MrChrisW commented on share/goodie/planets/planets.yml:
```diff
@@ -0,0 +1,154 @@
+---
+sun:
```
Remove the Sun? No image available, its also a star not planet.

mintsoft replied:
I'd say leave it in as it adds value; I'd "expect" sun results as well as planets. "SolarBodies.yml" ?

MrChrisW replied:
I agree :+1:

MrChrisW commented on t/Planets.t:
```diff
@@ -0,0 +1,63 @@
+
+   	DDG::Request->new(query_raw => "size of jupiter", location => test_location("au")) => test_zci("Jupiter, Radius is 69,911 km",  html => qr{<div class="zci--planets"><span class="planets--planetImage"><img src=".*" height="48" width="48"/></span><span class="planets--info"><span class="text--primary planets--planetAttribute">69,911 km</span><span class="text--secondary planets--planetName">Jup ...
+
+   	DDG::Request->new(query_raw => "size of planet jupiter", location => test_location("au")) => test_zci("Jupiter, Radius is 69,911 km",  html => qr{<div class="zci--planets"><span class="planets--planetImage"><img src=".*" height="48" width="48"/></span><span class="planets--info"><span class="text--primary planets--planetAttribute">69,911 km</span><span class="text--secondary planets--planetNa ...
```
Fix indentation.

MrChrisW commented on share/goodie/planets/planets.yml:
```diff
@@ -0,0 +1,154 @@
+  surface_area_imperial: 3,120,894,516 mi2
+  surface_area_scientific: 8.0831 x 109 km2
+  volume: 68,334,355,695,584 km3
+  volume_imperial: #not found
```
Fix this.

MrChrisW commented on share/goodie/planets/planets.yml:
```diff
@@ -0,0 +1,154 @@
+  density: 1.409 g/cm3
+mercury:
+  radius: 2,439.7 km
+  radius_imperial: 1,516.0 miles
```
Round or remove decimal point in all values?

mintsoft replied:
You could keep them in if they're significant; i.e. shows the level of measurement accuracy

MrChrisW commented:
@mintsoft As my other response got buried In an outdated diff I'll re-paste here.

So I'd like to clarify the best way to proceed:

 1. Switching units based on users preferences (not sure this is accessible without using JS)

 2. Switching units based on location (currently in use)

 3. Not switching units, all units as SI (modern metric)

jagtalon commented:
@MrChrisW holy crap this looks pretty.

jagtalon commented on lib/DDG/Goodie/SolarSystem.pm:
```diff
@@ -0,0 +1,113 @@
+  # Declare vars
+  my ($attribute, $attributesString, $result, $objectObj);
+  
+  s/what is (the)|(of)|(object)//g; # Remove common words, strip question marks
```
Should there be a `^` at the beginning of `what`?

jagtalon commented on lib/DDG/Goodie/SolarSystem.pm:
```diff
@@ -0,0 +1,113 @@
+  $html .= "</span>";
+  $html .= "<span class=\"objects--info\">";
+  $html .= "<span class=\"text--primary objects--objectAttribute\">";
+  $html .= $result;
```
Encode these?

MrChrisW replied:
Unfortunately that would result in an issue with using the `<sup></sup>` tags in [line 83](https://github.com/MrChrisW/zeroclickinfo-goodies/blob/060a2af098e4c000271813aadb16cb5783093247/lib/DDG/Goodie/SolarSystem.pm#L83) all values are pulled directly from the YML data, It's doubtful there's any XSS risk here.

jagtalon commented:
@abeyang I wonder how we can do the attribution with this one.

https://ddh1.duckduckgo.com/?q=what+is+the+size+of+venus&ia=answer
https://ddh1.duckduckgo.com/?q=size+of+earth&ia=answer

jagtalon commented:
With attribution I mean the person who made the images.

MrChrisW commented:
@jagtalon I was hoping @abeyang could make me some new sexy looking planets? :laughing:

jagtalon commented:
@MrChrisW oh that's another route.

abeyang commented:
Hi @MrChrisW , apologize for my silence... having some internal discussions re: planet images. Hope to get back to you soon!

jagtalon commented:
@MrChrisW I'm not sure how important it is, but I think it would be good if we could specify if its either mean or equatorial radius that we're displaying.

MrChrisW commented:
@jagtalon Sure, I think mean radius is assumed but it's probably best to include it.

abeyang commented:
Hey @MrChrisW, just wanna give you an update. We're thinking of possibly creating the icons ourselves, but we don't really get to it til sometime next week. Hope you're OK waiting until then...

MrChrisW commented:
Hey @abeyang no problem, looking forward to seeing your grand design concept  :smile:

abeyang commented:
Hey @MrChrisW , sorry for the delay. We had to make the rounds with @thm for these planet designs. We *may* need to update Saturn, but aside from that, the rest are ready to go!

In the end, it should look something like this:

![planets-v1 1-earth](https://cloud.githubusercontent.com/assets/794084/6068356/3923547c-ad30-11e4-9e6a-3d4330daa76b.jpg)

Note that Saturn has to push the text out a bit:

![planets-v1 1-saturn](https://cloud.githubusercontent.com/assets/794084/6068357/39241344-ad30-11e4-9c97-132d242e13bb.jpg)

And here are the assets:
http://cl.ly/2s2r3O2o3n1r

Can't wait to see this up and running!

mintsoft commented:
@abeyang @thm Those look awesome!

MrChrisW commented:
@abeyang Fantastic, thanks! Will implement shortly.

P.S Thanks @thm

MrChrisW commented:
Updated as per @abeyang feedback, let me know if the styles need adjusting. 

I think the `margin-right` on `objectImage` needs adjusting when the image is Saturn as it pushes the text to far to the right. http://duckpan1.mrchris.me/?q=size+of+saturn
 
Viewable here: http://duckpan1.mrchris.me/?q=size+of+earth

Still a few more issues I'm looking to clear up. 

+ Decide if this [conditianl statment](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/939/files#diff-cde95278264cf233c2de28ccf5ffa736R56) used to switch units from metric to imperial is required. Otherwise just display in metric. 
+ Fix [volume_imperial](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/939/files#diff-cdfd7be09ec7cf342db5db5fd676ddfeR129) for uranus
+ Double check all planet values are correct. 
+ Resolve style issue with Saturn.
+ Include Pluto data? as we have an image for it. (poort pluto :cry: )

MrChrisW commented:
Looks like some internal triggers will need to be removed if this is released #626

MrChrisW commented:
The CSS loading after the answer has been displayed really annoys me, is this only on DuckPan? 

![out](https://cloud.githubusercontent.com/assets/5282264/6081799/5191982a-ae57-11e4-9af6-26e69fd1af2f.gif)

moollaza commented:
@MrChrisW 

>Looks like some internal triggers will need to be removed if this is released #626

Yup but that's a quick fix ;)

>The CSS loading after the answer has been displayed really annoys me, is this only on DuckPan? 

I noticed that too. I'm not entirely sure, but I think we recently discovered a Goodie-Spice race condition that we're working to fix. I see the CSS only loads after the 'Images' and 'Videos' tabs appear.

------

Nasa has a TON of awesome data on these planets, http://solarsystem.nasa.gov/planets/profile.cfm?Object=Uranus&Display=Facts&System=Metric has the missing Uranus data I believe

>Include Pluto data?

Yes! It's technically a "dwarf planet" so I would argue that belongs to the "Planets" Goodie :)

MrChrisW commented:
@moollaza 

>I noticed that too. I'm not entirely sure, but I think we recently discovered a Goodie-Spice race condition that we're working to fix. I see the CSS only loads after the 'Images' and 'Videos' tabs appear.

Great, looking forward to seeing a fix in the future. 

>Nasa has a TON of awesome data on these planets,

That's my source! Shhh don't tell anyone.

>Yes! It's technically a "dwarf planet" so I would argue that belongs to the "Planets" Goodie :)

I made the choice to change the name to SolarSystem Goodie as techically the sun isn't a planet! :laughing: I'll add pluto pronto :+1:

MrChrisW commented:
This should be ready for a design re-review from @abeyang 

+ View online: [Earth](http://duckpan1.mrchris.me/?q=size+of+earth) | [Saturn](http://duckpan1.mrchris.me/?q=size+of+saturn)
+ [CSS](https://github.com/MrChrisW/zeroclickinfo-goodies/blob/planets/share/goodie/solar_system/solar_system.css)

@moollaza @jagtalon Still need an answer regarding the display of metric vs non-metric? See [comment](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/939#issuecomment-70767197)

mintsoft commented:
@MrChrisW Metric! Sod the yanks!

MrChrisW commented:
@mintsoft Amen to that! :laughing:

abeyang commented:
Sorry @MrChrisW -- had to deliberate on how to better show the planets. We finalized with this:

![planets-v2-earth](https://cloud.githubusercontent.com/assets/794084/6472936/ae6e856a-c1aa-11e4-9a28-dfe634b7e214.jpg)

![planets-v2-saturn](https://cloud.githubusercontent.com/assets/794084/6472938/ae836246-c1aa-11e4-865d-82bff5cd80b8.jpg)

The background cover is 60px, and each planet/sun/moon is 40px (all except Saturn). Here are the updated assets: http://cl.ly/0d2b1M0Q0E0i

Finally, can we try to vertically-center the text with the cover? Seems a bit misaligned right now. You can try to use the classnames proposed under **Images** in our [styleguide](https://duckduckgo.com/styleguide):

```html
<div class="zci__main">
	<div class="zci__img-wrap">
		<img class="zci__img">
	</div>
	<div class="zci__body">Text Goes Here</div>
</div>
```

Thanks! Can't wait to see this go live!

MrChrisW commented:
@abeyang Thanks for getting back with an update on this :smile: I'll update and push shortly!

moollaza commented:
@MrChrisW just checking in -- is this ready for us or are there more changes coming?

MrChrisW commented:
@moollaza This is still awaiting some small changes from **me** :smile: I'll ping when ready :+1:

MrChrisW commented:
@abeyang This has been updated :+1: 

+ View online: [Earth](http://duckpan1.mrchris.me/?q=size+of+earth) | [Saturn](http://duckpan1.mrchris.me/?q=size+of+saturn)

I had a go at using the CSS classes recommended in the styleguide (which looks amazing btw :astonished: ) however after a while I decided to fallback on the classes I'd already created as I couldn't get it looking quite right :frowning: 

Text should be vertically centred now:

![menu_295](https://cloud.githubusercontent.com/assets/5282264/6731193/15b78c6e-ce7e-11e4-914c-b3dc90652092.png)
![menu_296](https://cloud.githubusercontent.com/assets/5282264/6731194/15bb958e-ce7e-11e4-9745-c082e55df486.png)

mintsoft commented:
@MrChrisW Was just going to tell you about that ;)

MrChrisW commented:
@mintsoft You're totally the reason I added that edit! :laughing:

MrChrisW commented:
Any suggestions on dealing overflow of large numbers on mobile? We could show the scientific notation :grey_question: 

![selection_300](https://cloud.githubusercontent.com/assets/5282264/6765128/9a054246-d00d-11e4-82a0-55dc70a4ec88.png)

mintsoft commented:
@MrChrisW Definitely scientific notation. Or alternatively do it in zetta-lbs?

MrChrisW commented:
@mintsoft Thanks! I'm leaning towards scientific notation, I feel like we need to choose one way or the other not a mix of both. I'd appreciate any feedback from the design team or other staff when available :+1:

mintsoft commented:
Yeah, makes sense @abeyang or @chrismorast any opinion or thoughts?

abeyang commented:
Can we do something like this:
![planets-v2-saturn-mass](https://cloud.githubusercontent.com/assets/794084/6832219/4fdcdf44-d2e2-11e4-9e50-d318235cf3c9.jpg)

So, I know we also mentioned having the western notation, but thinking back to my high school days of unit conversions (way back in the day), I don't ever recall using pounds -- whereas miles (vs km) still seems to make some sense here.

In any case, the more important factor is showing the exponential factor; multiplication sign is `&times;` I believe.

MrChrisW commented:
@abeyang thanks. The scientific notations provided by NASA only include metric, I'm guessing I'll need to convert the values. See [example data](https://solarsystem.nasa.gov/planets/profile.cfm?Object=Mars&Display=Facts)

My main question is, should we display all values as scientific notation or only the few that are too long to be displayed in decimal notation?

abeyang commented:
@MrChrisW I don't think we need to use scientific notation for everything. Maybe a good basis (for now) for when to use scientific notation is when the value goes above one million...

MrChrisW commented:
@abeyang Updated again to reflect the required changes:

Examples:
http://duckpan1.mrchris.me/?q=mass+moon
http://duckpan1.mrchris.me/?q=volume+neptune
http://duckpan1.mrchris.me/?q=mass+of+earth

Note the differing number of decimal places on Neptune vs Moon, I'm unsure if we should round all these values to 3 decimal places. @mintsoft

mintsoft commented:
@MrChrisW I'd say round it for consistency; it looks a little weird having different sig-figs

MrChrisW commented:
![selection_304](https://cloud.githubusercontent.com/assets/5282264/6860969/88ffb1f0-d46a-11e4-992c-221999f6c9c8.png)

IE 8 planet vertical align is broken.

MrChrisW commented:
Fixed IE8 issue by removing the use of CSS3 `transform` and reverting to table-cell and vertical-align :frowning: rounded the scientific notations.

mintsoft commented:
@MrChrisW If the images were changed to be `div`s with a background image of the image that you are looking for then you could switch them out with media queries?

moollaza commented:
@MrChrisW @mintsoft we should be able to handle the img loading in the template, if you're interested in updating this to use the awesome new Goodie Templates?!

We'd need to use a custom template here, but that shouldn't cause any problems. Here's an example to follow @MrChrisW: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/BPMToMs.pm#L64

I swear actual documentation is coming soon :smile:

MrChrisW commented:
Thanks for the suggestions @mintsoft @moollaza 
I'll update to use a template, looks simple enough :wink: 
>I swear actual documentation is coming soon :smile:

I'll believe that when I see it ! :laughing:

mintsoft commented:
@moollaza Heard it all before ;)

MrChrisW commented:
@moollaza I've begun the process of converting the IA to use the new templates. 

The issue I'm running into when using the `base` template, it seems to want to return multiple results as tiles, see an example of my [structured_answer](https://gist.github.com/MrChrisW/b8313af838ae7331a3f8) and the result 
![selection_306](https://cloud.githubusercontent.com/assets/5282264/6880098/8a11ee2e-d559-11e4-9428-e9c2d12aabf1.png)

It's probably something silly I'm overlooking! Thanks

moollaza commented:
@MrChrisW it looks like you disabled the detail template? I believe that's causing the problem?

MrChrisW commented:
@moollaza :+1:

MrChrisW commented:
@moollaza Updated to use the Goodies template. [Working well](http://duckpan1.mrchris.me/?q=size+of+mars&ia=answer)

>We should be able to handle the img loading in the template

Further on this comment, I take it we simply want to send the planet name to the template and load the image using an `img` tag (easy!). 

I'm still confused as to how this will enable us to detect mobile and change the image accordingly?

MrChrisW commented on share/goodie/solar_system/solar_system.css:
```diff
@@ -0,0 +1,29 @@
+    display: table-cell;
+    vertical-align: middle;
+    background-color: #333;
+    border-radius: 5%;
```
Won't work with IE8 :+1:

mintsoft replied:
Should be fine; it'll degrade gracefully

MrChrisW commented:
@moollaza I've added a new mobile template. Using the `detail_mobile` option. See 
[mobile.handlebars](https://github.com/MrChrisW/zeroclickinfo-goodies/blob/planets/share/goodie/solar_system/mobile.handlebars), [content.handlebars](https://github.com/MrChrisW/zeroclickinfo-goodies/blob/planets/share/goodie/solar_system/content.handlebars)

I've moved the width & height values inline in order to set a differing height width if the image is Saturn. 

@abeyang I've noticed on mobile the `.zci` selector padding is modified to 0px. Can I set the padding here as it makes things look a little strange (see below)
```css
.zci {
    padding-left: 0px;
    padding-right: 0px;
}
```

![selection_313](https://cloud.githubusercontent.com/assets/5282264/6978099/1f0feb36-d9fe-11e4-9533-325b73bb0f8a.png)

mintsoft commented:
@MrChrisW Looks like there's a `html_enc` problem here:
http://duckpan1.mrchris.me/?q=mass+of+earth&ia=answer

MrChrisW commented:
@mintsoft Nice catch, that's what I get for not including full tests of returned `structured_data` :frowning: 

Fixed :tada:

mintsoft commented:
@MrChrisW Not sure what i've done here, I was just resizing the window a little and it's got stuck like this:
![image](https://cloud.githubusercontent.com/assets/978048/6987824/92bff57c-da43-11e4-8773-0897b8e82ca1.png)

moollaza commented:
@MrChrisW I spoke with @chrismorast and he says we should be trying to use SVG wherever possible (which means we'll need to get SVG assets from the design team wherever we're currently using PNG), which would really simplify image handling (because we won't need to check for retina devices.

For Goodies, if we have an SVG we can slurp the content directly into the HTML.

> am I linking to the images correctly? 

@MrChrisW when we link to images in production, the path are versioned, so you'll need to include the current Goodie version number. For now, you can get the current version number in Perl with `{{ENV{DDG_GoodieBundle_OpenSourceDuckDuckGo}}}`:

```handlebars
<img src="/share/goodie/{{goodie_version}}/{{img}}">
```

moollaza commented:
Once we have SVG you could slurp the SVG straight into the HTML. We could even Base64 encode them!

mintsoft commented:
@moollaza does the push towards SVG's mean we're sidelining IE8? :grinning:

moollaza commented:
@mintsoft damnit, that's why we're using PNG's. I couldn't remember the reason. Everywhere else on the site we fallback to PNG when SVG won't work (we use Modernizr, so `Modernizr.svg` tells us when we can use them). I'm not sure if we have a simple solution for that via the templates yet so I guess we'll need to stick to PNG for now.

MrChrisW commented:
@mintsoft Confirmed broken in Chrome! Thanks for testing :+1: 

@moollaza The IA was originally using SVG images and slurping them using `DDG::GoodieRole::ImageLoader` then the design team switched to PNG images :confused: I'll leave it as is for now...

Regarding the image paths, I'm assuming the correct path is like this ?
`<img src="/share/goodie/{{goodie_version}}/solar_system/img/{{imageName}}.png`

I'll leave this until the last second (commit) as this will break testing on DuckPAN :+1:

MrChrisW commented:
@mintsoft I think I've fixed the Chrome newline issue, can you [confirm](http://duckpan1.mrchris.me/?q=mass+of+earth&ia=answer)? Thanks :+1:

mintsoft commented:
@MrChrisW Yup, all fixed! 
![image](https://cloud.githubusercontent.com/assets/978048/6991955/2512dbe0-daaa-11e4-9f57-f6562bf28097.png)

MrChrisW commented:
Suggestion: Add the ability to switch units using a trigger like: 
+ Search Default = Location specific
+ Search `mass sun kg` = Metric
+ Search `mass sun lbs` = United States customary units

mintsoft commented:
That sounds cool @MrChrisW although I'd probably also want `mass sun metric` or `mass sun imperial`

moollaza commented:
> I'll leave this until the last second (commit) as this will break testing on DuckPAN :+1: 

What error are you getting from DuckPAN? I thought I fixed the subdirectory problem...

MrChrisW commented:
@moollaza 
>when we link to images in production, the path are versioned, so you'll need to include the current Goodie version number.

Including the versioned path in a DuckPAN environment won't work as the path is only versioned in prod? Right?

Secondly I cannot seem to use the environment var like this 
```perl
{{ENV{DDG_GoodieBundle_OpenSourceDuckDuckGo}}}
``` 

as it results in 
```perl 
Bareword "DDG_GoodieBundle_OpenSourceDuckDuckGo" not allowed while "strict subs"
```

MrChrisW commented:
@mintsoft Those types of queries should work now :+1: There's probably improvements that can be made, see a81199d

+ http://duckpan1.mrchris.me/?q=size+of+mars+metric&ia=answer
+ http://duckpan1.mrchris.me/?q=mass+sun+kg&ia=answer

mintsoft commented:
@MrChrisW an improvement that springs to mind is supporting `mass of mercury in kg` or `size of venus in metric`

MrChrisW commented:
@mintsoft :+1: these queries work now.

moollaza commented:
>Including the versioned path in a DuckPAN environment won't work as the path is only versioned in prod. Right?

@MrChrisW  DuckPAN will ignore the version number so versioned paths should work fine :)

>Secondly I cannot seem to use the environment var like this 

My bad, I totally wasn't thinking straight. We export that variable in Production so it won't be available in DuckPAN. The installed Goodie Package will have a version though that you can access locally:

```perl
my $version = $DDG::GoodieBundle::OpenSourceDuckDuckGo::VERSION // 0;
```

This will give you `0` when using DuckPAN (because it loads the files from the repo's local `/lib` instead of your `/perl ` install dir, but it should work for us in Production. I'll make a task to expose the Goodie version to the JS so you can access it more easily from the templates (we do this already for `spice_version` but haven't needed it yet for Goodies).

MrChrisW commented:
@moollaza No problem! Thanks.

>DuckPAN will ignore the version number so versioned paths should work fine

Doesn't look like that's the case as accessing `/share/goodie/0/solar_system/img/mars.png` fails with `ERROR: File not found`

Example: 
+ http://duckpan1.mrchris.me/share/goodie/solar_system/img/mars.png 
+ http://duckpan1.mrchris.me/share/goodie/0/solar_system/img/mars.png

Looks like it may have something to do with commit [021f3a2bc7792b83eef5512d17860510ba6c1b08](https://github.com/duckduckgo/p5-app-duckpan/commit/021f3a2bc7792b83eef5512d17860510ba6c1b08#diff-ebc5e1818c64eceac5960f3997f6aca9R98)

abeyang commented:
@MrChrisW looking good! Can we make these minor font adjustments:

``` css
.zci--solar_system .objects--objectAttribute {
  font-size: 1.319em;
}

.zci--solar_system .objects--objectName {
  font-size: 0.972em;
}
```

Almost there!

moollaza commented:
>Doesn't look like that's the case as accessing /share/goodie/0/solar_system/img/mars.png fails with ERROR: File not found

@MrChrisW mind changing `0` to `000` or `999` for now? It's a hack but it'll get things moving ahead for you. I've made a task for better image loading for Goodies. DuckPAN only expects `\d{3,4}` because that's where we're at for Spice versions. I feel like there was a reason why I didn't just generalize it to `\d+` but I can't think of it right now.

MrChrisW commented:
@moollaza I did see that line `\d{3,4}` and tried accessing using 3-4 numbers however this still fails with the same result:

Test: http://duckpan1.mrchris.me/share/goodie/000/solar_system/img/mars.png

Thanks

MrChrisW commented:
@abeyang Hello! Thanks will update with this adjustments, can you comment on the `padding` value on mobile? Thanks

moollaza commented:
@MrChrisW I'm an idiot -- made a typo in the share path structure!

http://duckpan1.mrchris.me/share/goodie/solar_system/000/img/mars.png

The version goes after the IA name.

MrChrisW commented:
@moollaza Awesome! I'll update shortly. Glad we finally got to the bottom of that :laughing:

mintsoft commented:
@MrChrisW are you just waiting on feedback from @abeyang now?

MrChrisW commented:
@mintsoft That's correct. 

In the meantime feel free to nitpick at any of the code :smile:

mintsoft commented on lib/DDG/Goodie/SolarSystem.pm:
```diff
@@ -0,0 +1,133 @@
+my $objects = Load(scalar share('objects.yml')->slurp);
+
+# Handle statement
+handle query_lc => sub {
```
@MrChrisW Speaking of nit-picking; this seems to be indented to 2 spaces when it should be 4!

mintsoft commented on lib/DDG/Goodie/SolarSystem.pm:
```diff
@@ -0,0 +1,133 @@
+  $saturn = ($objectName eq "saturn") ? 1 : 0;
+
+  #Return result and html
+      return $operation." is ".$result,
```
Also odd indentation here

mintsoft commented on lib/DDG/Goodie/SolarSystem.pm:
```diff
@@ -0,0 +1,133 @@
+
+  # Superscript for km3, mi3, km2 or mi2 
+  if($result =~ m/(km|mi)(\d)/) {
+    my $symbol = $1; my $superscript = $2;
```
You could do
```perl
my ($symbol, $superscript) = ($1, $2);
```
if you wanted?

mintsoft commented on share/goodie/solar_system/content.handlebars:
```diff
@@ -0,0 +1,19 @@
+<div class="objects--body">
+    <span class="objects--objectImage">
+        {{#if saturn}}    
+            <img width="45px" height="28px" src="share/goodie/solar_system//{{goodie_version}}/img/{{imageName}}.png">
```
Hmm I think if the width and height aren't set then it'll default to the resolution of the png? So we can bomb the if and the special "is saturn" flag too?

mintsoft commented on share/goodie/solar_system/mobile.handlebars:
```diff
@@ -0,0 +1,19 @@
+<div class="objects--body">
+    <span class="objects--objectImage">
+        {{#if saturn}}    
+            <img width="45px" height="28px" src="share/goodie/solar_system//{{goodie_version}}/img/{{imageName}}@2x.png">
```
Same again, although I'm guessing the sizes are here for increased screen pixels? Can we set them to 50% to get the same effect instead?

MrChrisW replied:
Regarding the inline styles, we can remove the `width` and `height` values for the main `content.handlebars` (desktop) and the images will look normal. 

However with the `mobile.handlebars` as the images are double the size `width` and `height` need to be set, unfortunately setting these values to `50%` will only set them image to half the size of it's [surrounding container](https://github.com/MrChrisW/zeroclickinfo-goodies/blob/planets/share/goodie/solar_system/solar_system.css#L2-L3) (60px, 60px)

MrChrisW commented:
@mintsoft Thanks for nitpicking! :smile:

abeyang commented:
@MrChrisW looks good! One **final** nitpick -- can we have the surface area return in scientific notation as well? Any reason why it's not?

![screen shot 2015-04-13 at 12 28 44 pm](https://cloud.githubusercontent.com/assets/794084/7123668/ab30664a-e1d8-11e4-934f-73f73cfa5821.png)

chrismorast commented:
Are you sure @abeyang ? I much prefer to see the number written out over the scientific notation

MrChrisW commented:
@abeyang Any numbers that overflow on a screen 350px> are scientific notation. Sun Surface Area does not :smile: 
![selection_345](https://cloud.githubusercontent.com/assets/5282264/7130280/f66978b4-e29f-11e4-95c5-420d797faee5.png)

@chrismorast @abeyang Are you able to recommend a [fix for the `padding-left`](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/939#issuecomment-89159354) on mobile?

abeyang commented:
Hey @MrChrisW , I'm sorry, but my dev server decided to stop working for me. Can you throw it onto yours for me? Thanks so much!

abeyang commented:
For the mobile, let's try this:

``` css
@media only screen and (max-width: 44em) {
  .zci {
    padding-left: 1.2em;
  }
}
```

@sdougbrown @moollaza Please LMK if you have better suggestions...

MrChrisW commented:
@abeyang Thanks, looks like my server decided to kill DuckPAN too :laughing: [Fixed now](http://duckpan1.mrchris.me/?q=mass+of+the+sun&ia=answer)

I'll update with the new CSS once @moollaza approves :smile: I wonder if there's any other IAs that push the padding-left on mobile?

MrChrisW commented:
Ping @moollaza

moollaza commented:
@MrChrisW sorry, that particular CSS won't work because it'll affect any/all other IA's when this Solar System IA CSS loads -- I'll take a peak today to figure it out!

moollaza commented:
@MrChrisW looks like the `<div class="object--body">` should have the class `zci__body` -- that'll add the correct side padding:

![2015-04-17_11h17_59](https://cloud.githubusercontent.com/assets/873785/7204879/76c3827a-e4f3-11e4-9afc-85c75e5c4ade.png)

That may be a missing component of the template though -- I'll double check.

moollaza commented:
The problem is that the mobile_detail isn't getting wrapped in our `base_detail` template. If you specify `wrap_detail => 'base_detail', it fixes the mobile view, but double-wraps the normal, detail view, which is a bug. I'll make a task to fix this.

/cc @sdougbrown

MrChrisW commented:
@moollaza Thanks for looking into that.

Should I proceed to update, specifying `wrap_detail` and await the internal fix or leave as is?

moollaza commented:
@MrChrisW I think the best way forward is to add `zci--body` to your mobile template for now. We'll need to coordinate the internal fix with removing that to prevent double padding though

MrChrisW commented:
@moollaza Sure, I've added `zci__body` to the mobile template.

http://duckpan1.mrchris.me/?q=size+of+mars+metric&ia=answer

moollaza commented:
@MrChrisW :+1: 

@abeyang I believe this is ready for a final design review? The mobile padding has been fixed. If that's all that was needed please merge :+1:

abeyang commented:
:boom:

mintsoft commented:
Awesome; can't wait to see this live @MrChrisW !

MrChrisW commented:
Thanks for the great feedback everyone :+1:  @mintsoft ! @moollaza @abeyang @jagtalon @chrismorast 
___________________![](http://i.imgur.com/q2iS18y.gif)__________________

jagtalon commented:
![01n4njx](https://cloud.githubusercontent.com/assets/81969/7438349/d74b6930-f030-11e4-8961-f2783a42c9af.gif)

moollaza commented:
@MrChrisW sorry for the long delay, I hadn't realized this was merged and ready for release. Will set it up to go live with the next set of IAs!

MrChrisW commented:
@moollaza I was starting to wonder.. :wink: Thanks for the update!

moollaza commented:
@MrChrisW IT'S LIVE!

https://duckduckgo.com/?q=mass+of+earth&t=ffab&ia=answer

Sorry about the long delay. Thanks for another awesome IA :+1:

MrChrisW commented:
@moollaza AMAZING!! Thanks
